File: Sources/MirageKit/Internal/Protocol/MessageTypes.swift
//
//  MessageTypes.swift
//  MirageKit
//
//  Created by Ethan Lipnik on 1/2/26.
//

import CoreGraphics
import Foundation

/// Control channel message types (sent over TCP)
enum ControlMessageType: UInt8, Codable {
    // Connection management
    case hello = 0x01
    case helloResponse = 0x02
    case disconnect = 0x03
    case ping = 0x04
    case pong = 0x05

    // Authentication
    case authRequest = 0x10
    case authChallenge = 0x11
    case authResponse = 0x12
    case authResult = 0x13

    // Window management
    case windowListRequest = 0x20
    case windowList = 0x21
    case windowUpdate = 0x22
    case startStream = 0x23
    case stopStream = 0x24
    case streamStarted = 0x25
    case streamStopped = 0x26
    case streamMetricsUpdate = 0x27

    /// Input events
    case inputEvent = 0x30

    /// Keyframe control
    case keyframeRequest = 0x42

    /// Cursor updates
    case cursorUpdate = 0x50
    case cursorPositionUpdate = 0x51

    // Virtual display updates
    case contentBoundsUpdate = 0x60
    case displayResolutionChange = 0x61
    case streamScaleChange = 0x62
    case streamRefreshRateChange = 0x63

    // Session state and unlock (for headless Mac support)
    case sessionStateUpdate = 0x70
    case unlockRequest = 0x71
    case unlockResponse = 0x72
    case loginDisplayReady = 0x73 // Host -> Client: Login display stream is starting
    case loginDisplayStopped = 0x74 // Host -> Client: Login complete, display stream stopped

    // App-centric streaming (new)
    case appListRequest = 0x80
    case appList = 0x81
    case selectApp = 0x82
    case appStreamStarted = 0x83
    case windowAddedToStream = 0x84
    case windowRemovedFromStream = 0x85
    case windowCooldownStarted = 0x86
    case windowCooldownCancelled = 0x87
    case returnToAppSelection = 0x88
    case closeWindowRequest = 0x89
    case streamPaused = 0x8A
    case streamResumed = 0x8B
    case cancelCooldown = 0x8C
    case windowResizabilityChanged = 0x8D
    case appTerminated = 0x8E

    // Menu bar passthrough
    case menuBarUpdate = 0x90 // Host → Client: Menu structure update
    case menuActionRequest = 0x91 // Client → Host: Execute menu action
    case menuActionResult = 0x92 // Host → Client: Action result

    // Desktop streaming (full virtual display mirroring)
    case startDesktopStream = 0xA0 // Client → Host: Start full desktop stream
    case stopDesktopStream = 0xA1 // Client → Host: Stop desktop stream
    case desktopStreamStarted = 0xA2 // Host → Client: Desktop stream is active
    case desktopStreamStopped = 0xA3 // Host → Client: Desktop stream ended
    case qualityTestRequest = 0xA4 // Client → Host: Run quality test
    case qualityTestResult = 0xA5 // Host → Client: Quality test metadata/result

    /// Errors
    case error = 0xFF
}

/// Base control message envelope
struct ControlMessage: Codable {
    let type: ControlMessageType
    let payload: Data

    init(type: ControlMessageType, payload: Data = Data()) {
        self.type = type
        self.payload = payload
    }

    init(type: ControlMessageType, content: some Encodable) throws {
        self.type = type
        payload = try JSONEncoder().encode(content)
    }

    func decode<T: Decodable>(_ type: T.Type) throws -> T {
        try JSONDecoder().decode(type, from: payload)
    }

    func serialize() -> Data {
        var data = Data()
        data.append(type.rawValue)
        withUnsafeBytes(of: UInt32(payload.count).littleEndian) { data.append(contentsOf: $0) }
        data.append(payload)
        return data
    }

    static func deserialize(from data: Data) -> (ControlMessage, Int)? {
        guard data.count >= 5 else { return nil }

        // Use index-relative access for Data that might be a slice
        let startIdx = data.startIndex
        let typeByte = data[startIdx]
        guard let type = ControlMessageType(rawValue: typeByte) else {
            MirageLogger.error(.client, "Unknown control message type byte: 0x\(String(format: "%02X", typeByte))")
            return nil
        }

        // Read length from bytes 1-4 (after the type byte)
        let lengthBytes = data[data.index(startIdx, offsetBy: 1) ..< data.index(startIdx, offsetBy: 5)]
        let length = lengthBytes.withUnsafeBytes { ptr in
            UInt32(littleEndian: ptr.loadUnaligned(fromByteOffset: 0, as: UInt32.self))
        }

        let totalLength = 5 + Int(length)
        guard data.count >= totalLength else { return nil }

        // Extract payload using proper indices
        let payloadStart = data.index(startIdx, offsetBy: 5)
        let payloadEnd = data.index(startIdx, offsetBy: totalLength)
        let payload = Data(data[payloadStart ..< payloadEnd])

        return (ControlMessage(type: type, payload: payload), totalLength)
    }
}

File: android/miragekit/src/main/kotlin/com/miragekit/protocol/MessageTypes.kt
package com.miragekit.protocol

import kotlinx.serialization.Serializable

@Serializable
enum class ControlMessageType(val value: UByte) {
    // Connection management
    HELLO(0x01u),
    HELLO_RESPONSE(0x02u),
    DISCONNECT(0x03u),
    PING(0x04u),
    PONG(0x05u),

    // Authentication
    AUTH_REQUEST(0x10u),
    AUTH_CHALLENGE(0x11u),
    AUTH_RESPONSE(0x12u),
    AUTH_RESULT(0x13u),

    // Window management
    WINDOW_LIST_REQUEST(0x20u),
    WINDOW_LIST(0x21u),
    WINDOW_UPDATE(0x22u),
    START_STREAM(0x23u),
    STOP_STREAM(0x24u),
    STREAM_STARTED(0x25u),
    STREAM_STOPPED(0x26u),
    STREAM_METRICS_UPDATE(0x27u),

    // Input events
    INPUT_EVENT(0x30u),

    // Keyframe control
    KEYFRAME_REQUEST(0x42u),

    // Cursor updates
    CURSOR_UPDATE(0x50u),
    CURSOR_POSITION_UPDATE(0x51u),

    // Virtual display updates
    CONTENT_BOUNDS_UPDATE(0x60u),
    DISPLAY_RESOLUTION_CHANGE(0x61u),
    STREAM_SCALE_CHANGE(0x62u),
    STREAM_REFRESH_RATE_CHANGE(0x63u),

    // Session state
    SESSION_STATE_UPDATE(0x70u),
    UNLOCK_REQUEST(0x71u),
    UNLOCK_RESPONSE(0x72u),
    LOGIN_DISPLAY_READY(0x73u),
    LOGIN_DISPLAY_STOPPED(0x74u),

    // App-centric streaming
    APP_LIST_REQUEST(0x80u),
    APP_LIST(0x81u),
    SELECT_APP(0x82u),
    APP_STREAM_STARTED(0x83u),
    WINDOW_ADDED_TO_STREAM(0x84u),
    WINDOW_REMOVED_FROM_STREAM(0x85u),
    WINDOW_COOLDOWN_STARTED(0x86u),
    WINDOW_COOLDOWN_CANCELLED(0x87u),
    RETURN_TO_APP_SELECTION(0x88u),
    CLOSE_WINDOW_REQUEST(0x89u),
    STREAM_PAUSED(0x8Au),
    STREAM_RESUMED(0x8Bu),
    CANCEL_COOLDOWN(0x8Cu),
    WINDOW_RESIZABILITY_CHANGED(0x8Du),
    APP_TERMINATED(0x8Eu),

    // Menu bar
    MENU_BAR_UPDATE(0x90u),
    MENU_ACTION_REQUEST(0x91u),
    MENU_ACTION_RESULT(0x92u),

    // Desktop streaming
    START_DESKTOP_STREAM(0xA0u),
    STOP_DESKTOP_STREAM(0xA1u),
    DESKTOP_STREAM_STARTED(0xA2u),
    DESKTOP_STREAM_STOPPED(0xA3u),
    QUALITY_TEST_REQUEST(0xA4u),
    QUALITY_TEST_RESULT(0xA5u),
    QUALITY_PROBE_REQUEST(0xA6u),
    QUALITY_PROBE_RESULT(0xA7u),

    ERROR(0xFFu);

    companion object {
        fun fromValue(value: UByte): ControlMessageType? {
            return entries.find { it.value == value }
        }
    }
}

data class ControlMessage(
    val type: ControlMessageType,
    val payload: ByteArray
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ControlMessage

        if (type != other.type) return false
        if (!payload.contentEquals(other.payload)) return false

        return true
    }

    override fun hashCode(): Int {
        var result = type.hashCode()
        result = 31 * result + payload.contentHashCode()
        return result
    }
}
